deps/acore/joiner/joiner.sh

```diff
@@ -0,0 +1,441 @@
+#!/usr/bin/env bash
+
+#
+# bash >= 4.x required
+#
+
+#
+# DEFINES
+#
+
+# boolean bash convention ( inverse )
+declare -A J_OPT;
+
+TRUE=0
+FALSE=1
+
+J_PATH="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
+
+unamestr=`uname`
+if [ -z "$J_PATH_MODULES" ]; then
+    if [[ "$unamestr" == 'Darwin' ]]; then
+        J_PATH_MODULES=$(greadlink -f "$J_PATH/../../")
+    else
+        J_PATH_MODULES=$(readlink -f "$J_PATH/../../")
+    fi
+fi
+
+for i in "$@"
+do
+case $i in
+    --parent=*) #internally used
+        J_OPT[parent]="${i#*=}"
+        shift
+    ;;
+    --child=*) #internally used
+        J_OPT[child]="${i#*=}"
+        shift
+    ;;
+    *)
+        # unknown option
+    ;;
+esac
+done
+
+J_PARAMS="$@"
+
+function Joiner:is_submodule()
+{
+    path=$1
+    (cd "$path" && cd "$(git rev-parse --show-toplevel 2>&1)/.."
+    git rev-parse --is-inside-work-tree 2>&1) | grep -q true
+}
+
+
+function Joiner:_help() {
+    hasReq=$1
+    firstParam=$2
+    msg=$3
+
+    if [ $hasReq = false ]; then
+        echo "Argument missing: $msg"
+        exit 1
+    fi
+
+    if [[ "$firstParam" = "--help" || "$firstParam" = "-h" ]]; then
+        echo "Help: $msg"
+        exit 1
+    fi
+}
+
+function Joiner:_searchFirstValiPath() {
+    path=$1
+    until $(cd -- "$path")
+    do
+        case   $path  in(*[!/]/*)
+            path="${path%/*}"
+        ;;
+        (*)
+            ! break
+    esac
+    done  2>/dev/null
+    echo $path
+}
+
+#
+# JOINER FUNCTIONS
+#
+
+function Joiner:add_repo() (
+    set -e
+    url=$1
+    name=${2:-""}
+    branch=${3:-"master"}
+    basedir=${4:-""}
+
+    [[ -z $url ]] && hasReq=false || hasReq=true
+    Joiner:_help $hasReq "$1" "Syntax: joiner.sh add-repo [-d] [-e] url name branch [basedir]"
+
+    # retrieving info from url if not set
+    if [[ -z $name ]]; then
+        basename=$(basename $url)
+        name=${basename%%.*}
+
+        if [[ -z $basedir ]]; then
+            dir=$(dirname $url)
+            basedir=$(basename $dir)
```

Comment:
I am wondering if the directory path might need to be quoted. Saw someone report the other day in discord that they got an error trying to install a module via acore.sh when they had a space in the directory path.

Reply:
It's out of scope since the joiner is an external tool (but maintained by us), however, I just applied your suggestion :) thanks